MDEV-33967 hander socket writes do not result in binary log entries

grooverdan opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling in this template <3

If you have any questions related to MariaDB or you just want to hang out and meet other community members, please join us on https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-33967*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
3. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description

As handlersocket bypasses the sql layer that prepares and causes binary log entries to be created, we need to do this in the plugin directly to create binary log entries.

Reported thanks to Yutaro Iwasaki on Zulip.

## Release Notes

Correct handler socket to allow changed rows to be replicated.

## How can this PR be tested?

Manual testing.

I attempted to restructure the tests, but cmake on top of two levels of build system for the perl module wasn't something I managed to unravel.

<!--
Tick one of the following boxes [x] to help us understand if the base branch for the PR is correct.
see [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) for the latest versions.
-->
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature or a refactoring, and the PR is based against the latest MariaDB development branch.*
- [X] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

<!--
  All code merged into the MariaDB codebase must meet a quality standard and codying style.
  Maintainers are happy to point out inconsistencies but in order to speed up the review and merge process we ask you to check the CODING standards.
-->
## PR quality check
- [X] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [X] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

grooverdan committed:
MDEV-33967 hander socket writes do not result in binary log entries

As handlersocket bypasses the sql layer that prepares and causes
binary log entries to be created, we need to do this in the
plugin directly to create binary log entries.

Reported thanks to Yutaro Iwasaki on Zulip.

Reviewer: Brandon Nesterenko

bnestere reviewed (changes requested):
Hi @grooverdan ! 

Thanks for the fix, it has been interesting getting to know the handlersocket plugin.

I have a few suggestions (not quite related to your fixes, but to complement them).

Also to the MTR test, I've manually created one by adding handlersocket parameters to a `.cnf` file, and using the handlersocket perl library to execute commands. Though to get that working generally would likely take more effort to generalize the ports in the MTR setup program, and find them in the perl script. FWIW, here is what I've been using for manual testing:

.cnf file:
```
!include ../my.cnf

[mysqld.1]
loose-handlersocket-address="127.0.0.1"
loose-handlersocket-port="9998"
loose-handlersocket-port_wr="9999"
log-slave-updates
loose-innodb

[mysqld.2]
log-slave-updates
loose-innodb
```

.test file:
```
--source include/have_innodb.inc
--source include/have_log_bin.inc
--source include/have_binlog_format_row.inc
--source include/master-slave.inc

--connection master
INSTALL PLUGIN handlersocket SONAME 'handlersocket.so';

--perl
use strict;
use warnings;

use DBI;
use Net::HandlerSocket;

my $port= $ENV{'MASTER_MYPORT'};
my $hsport= "9999";

my $dsn = "DBI:mysql:database=test;host=127.0.0.1;port=$port;mariadb_server_prepare=0";
my $dbh = DBI->connect($dsn, "root", "", { RaiseError => 1 });
my $hsargs = { 'host' => "127.0.0.1", 'port' => $hsport };
my $cli = new Net::HandlerSocket($hsargs);

$dbh->do("create table test.t1 (k int primary key, v int) engine=innodb");

$cli->open_index(1, "test", "t1", '', 'k,v');
$cli->execute_insert(1, [ 0, 0 ]);
$cli->execute_insert(1, [ 1, 1 ]);
$cli->execute_update(1, '=', [ 0 ], 1, 0, [ 0, 11 ]);
$cli->execute_delete(1, '=', [ 0 ], 1, 0);
EOF

--echo # Row 1,1 should be all that exists
select * from test.t1;


--echo #
--echo # Cleanup
--connection master
drop table t1;
--source include/save_master_gtid.inc

--connection slave
--source include/sync_with_master_gtid.inc

--source include/rpl_end.inc
--echo # End of rpl_handlersocket.test
```

grooverdan commented on plugin/handler_socket/handlersocket/database.cpp:
```diff
@@ -552,6 +552,8 @@ dbcontext::modify_record(dbcallback_i& cb, TABLE *const table,
     /* update */
     handler *const hnd = table->file;
+    table->mark_columns_needed_for_update();
+    hnd->prepare_for_row_logging();
```
> We always want `row` format when using the handlersocket, right? Even if the server is configured with `statement`?

yes. As mentioned constructing a statement seems overkill.

> I wonder if we should warn in such a case (to at least suggest configuring with `mixed` format),

I looked at the protocol and there's no extension for a warning - https://github.com/DeNA/HandlerSocket-Plugin-for-MySQL/blob/master/docs-en/protocol.en.txt

With MDEV-7389 however grabbing warnings and putting them in an audit log is possible. So worth while supporting.

> or perhaps try and construct a matching query to log in `statement` format (though that seems fickle, and more worth than it would be work).

It does a bit. There a case where event DML of tables is changed to statement without a warning so not totally without precedence to ignore the binlog format setting.

bnestere committed:
MDEV-33967 handlersocket initialize for binary logging

The thread is misconfigured for binlogging. When looking
at the binlog results, the event timestamp and xid is wrong.

This corrects by creating a time/xid for the binlog events.

grooverdan commented on plugin/handler_socket/handlersocket/database.cpp:
```diff
@@ -576,6 +578,8 @@ dbcontext::modify_record(dbcallback_i& cb, TABLE *const table,
     /* delete */
     handler *const hnd = table->file;
+    table->mark_columns_needed_for_delete();
+    hnd->prepare_for_row_logging();
```
Does this only apply to write ops?

Unless there are exposes in some information_schema table I don't think time/query_id are accessible?

So?
```
diff --git a/plugin/handler_socket/handlersocket/database.cpp b/plugin/handler_socket/handlersocket/database.cpp
index a144342b3f2..6463917f86a 100644
--- a/plugin/handler_socket/handlersocket/database.cpp
+++ b/plugin/handler_socket/handlersocket/database.cpp
@@ -1128,7 +1128,6 @@ dbcontext::cmd_exec(dbcallback_i& cb, const cmd_exec_args& args)
   }
   ha_rkey_function find_flag = HA_READ_KEY_EXACT;
   db_write_op wrop = db_write_op_none;
-  init_thd_for_query();
   if (args.op.size() == 1) {
     switch (args.op.begin()[0]) {
     case '=':
@@ -1170,8 +1169,10 @@ dbcontext::cmd_exec(dbcallback_i& cb, const cmd_exec_args& args)
   case db_write_op_none:
     return cmd_find_internal(cb, p, find_flag, args);
   case db_write_op_insert:
+    init_thd_for_query();
     return cmd_insert_internal(cb, p, args.kvals, args.kvalslen);
   case db_write_op_sql:
+    init_thd_for_query();
     return cmd_sql_internal(cb, p, args.kvals, args.kvalslen);
   }
 }
 ```

grooverdan committed:
MDEV-33967 handlersocket initialize for binary logging

Initialize for writing, and not for reading. The limited
range of hander socket read commands doesn't give easy
access to query_id and time information.

bnestere committed:
MDEV-33967 handlersocket correct memory leaks

There's a couple memory leaks in handersocket plugin:

1. The thread is not un-linked at destruction
2. The MDL keyname (handlersocket_wr) in lck_key is realloc-ed (by c_str_safe() when accessing the value) but never freed.

Note that the constructor to lck_key is changed to force a string copy, rather than directly pointing to our supplied handlersocket_wr.

grooverdan commented:
> Hi @grooverdan !
> 
> Thanks for the fix, it has been interesting getting to know the handlersocket plugin.
> 
> I have a few suggestions (not quite related to your fixes, but to complement them).

:-)

> 
> Also to the MTR test, I've manually created one by adding handlersocket parameters to a `.cnf` file, and using the handlersocket perl library to execute commands. Though to get that working generally would likely take more effort to generalize the ports in the MTR setup program, and find them in the perl script. FWIW, here is what I've been using for manual testing:

Thanks.

I managed to get some dynamic testing if port was free, but with 6 ports allocated for galera there should be plenty in the range for the worker. I was still unable to load shared libraries into the perl path.

https://github.com/MariaDB/server/compare/bb-10.5-MDEV-33967-handersocket-binlog-pkgtest...grooverdan:mariadb-server:bb-10.5-MDEV-33967-handersocket-binlog-pkgtest?expand=1

grooverdan commented:
updated, have one ASAN bug with this currently using test case provided:

```
#0  0x00007fbd272a8664 in __pthread_kill_implementation () from /lib64/libc.so.6
#1  0x00000000010ad04a in handle_fatal_signal (sig=<optimized out>) at /home/dan/repos/mariadb-server-10.5/sql/signal_handler.cc:365
#2  <signal handler called>
#3  0x00007fbd272a8664 in __pthread_kill_implementation () from /lib64/libc.so.6
#4  0x00007fbd2724fc4e in raise () from /lib64/libc.so.6
#5  0x00007fbd27237902 in abort () from /lib64/libc.so.6
#6  0x00000000007fa60b in __sanitizer::Abort() ()
#7  0x00000000007f87a5 in __sanitizer::Die() ()
#8  0x00000000007dc8bf in __asan::ScopedInErrorReport::~ScopedInErrorReport() ()
#9  0x00000000007db865 in __asan::ReportGenericError(unsigned long, unsigned long, unsigned long, unsigned long, bool, unsigned long, unsigned int, bool) ()
#10 0x00000000007dca6c in __asan_report_load1 ()
#11 0x0000000001953f7e in calc_row_difference (uvect=0x520000070cc8, old_row=0x519000221ad8 "", new_row=0x519000221ac8 "", table=0x519000221598, 
    upd_buff=0x50d000020ac8 "\200", buff_len=110, prebuilt=0x520000070108, auto_inc=<optimized out>)
    at /home/dan/repos/mariadb-server-10.5/storage/innobase/handler/ha_innodb.cc:8329
#12 ha_innobase::update_row (this=0x51d00010feb8, old_row=<optimized out>, new_row=<optimized out>)
    at /home/dan/repos/mariadb-server-10.5/storage/innobase/handler/ha_innodb.cc:8479
#13 0x00000000010d72c3 in handler::ha_update_row (this=0x51d00010feb8, old_data=<optimized out>, new_data=<optimized out>)
    at /home/dan/repos/mariadb-server-10.5/sql/handler.cc:7433
#14 0x00007fbd1cbcc1c9 in dena::dbcontext::modify_record (this=this@entry=0x50e000081140, cb=<optimized out>, table=table@entry=0x519000221598, 
    pst=@0x50d000020888: {dbctx = 0x50e000081140, table_id = 0, idxnum = 0, ret_fields = std::vector of length 2, capacity 2 = {0, 1}, filter_fields = std::vector of length 0, capacity 0}, 
    args=@0x7fbceab14060: {pst = 0x50d000020888, op = {start = 0x5210000dd502 "=\t1\t0\t1\t0\tU\t0\t11\n", '\276' <repeats 183 times>..., length = 1}, kvals = 0x7fbcebbff840, kvalslen = 1, limit = 1, skip = 0, mod_op = {start = 0x5210000dd50c "U\t0\t11\n", '\276' <repeats 193 times>..., length = 1}, uvals = 0x7fbcebbff7e0, filters = 0x0, invalues_keypart = -1, invalues = 0x0, invalueslen = 0}, mod_op=<optimized out>, modified_count=@0x7fbcea80a920: 0)
    at /home/dan/repos/mariadb-server-10.5/plugin/handler_socket/handlersocket/database.cpp:593
#15 0x00007fbd1cbce04b in dena::dbcontext::cmd_find_internal (this=<optimized out>, cb=<optimized out>, pst=<optimized out>, find_flag=<optimized out>, args=<optimized out>)
    at /home/dan/repos/mariadb-server-10.5/plugin/handler_socket/handlersocket/database.cpp:866
#16 0x00007fbd1cbe33ce in dena::hstcpsvr_worker::do_exec_on_index (this=<optimized out>, cmd_begin=<optimized out>, cmd_end=<optimized out>, 
    start=0x5210000dd512 "\n", '\276' <repeats 199 times>..., finish=<optimized out>, 
    conn=@0x5130000a9fc0: {<dena::dbcallback_i> = {_vptr$dbcallback_i = 0x7fbd1cbfa910 <vtable for dena::hstcpsvr_conn+16>}, fd = {<dena::noncopyable> = {<No data fields>}, fd = 89}, addr = {ss_family = 2, __ss_padding = "\252p\177\000\000\001", '\000' <repeats 111 times>, __ss_align = 0}, addr_len = 16, cstate = {readbuf = {<dena::noncopyable> = {<No data fields>}, buffer = 0x5210000dd500 "1\t=\t1\t0\t1\t0\tU\t0\t11\n", '\276' <repeats 181 times>..., begin_offset = 0, end_offset = 19, alloc_size = 4096}, writebuf = {<dena::noncopyable> = {<No data fields>}, buffer = 0x50300006e920 "0\t1\n", '\276' <repeats 28 times>, "\003\021", begin_offset = 0, end_offset = 0, alloc_size = 32}, prep_stmts = std::vector of length 2, capacity 2 = {{dbctx = 0x0, table_id = 18446744073709551615, idxnum = 18446744073709551615, ret_fields = std::vector of length 0, capacity 0, filter_fields = std::vector of length 0, capacity 0}, {dbctx = 0x50e000081140, table_id = 0, idxnum = 0, ret_fields = std::vector of length 2, capacity 2 = {0, 1}, filter_fields = std::vector of length 0, capacity 0}}, resp_begin_pos = 0, find_nl_pos = 0}, err = "", readsize = 0, nonblocking = true, read_finished = false, write_finished = false, nb_last_io = 1729081330, conns_iter = 0x5130000a9fc0, authorized = false}) at /home/dan/repos/mariadb-server-10.5/plugin/handler_socket/handlersocket/hstcpsvr_worker.cpp:883
#17 0x00007fbd1cbe1711 in dena::hstcpsvr_worker::execute_lines (this=this@entry=0x510000009040, 
    conn=@0x5130000a9fc0: {<dena::dbcallback_i> = {_vptr$dbcallback_i = 0x7fbd1cbfa910 <vtable for dena::hstcpsvr_conn+16>}, fd = {<dena::noncopyable> = {<No data fields>}, fd = 89}, addr = {ss_family = 2, __ss_padding = "\252p\177\000\000\001", '\000' <repeats 111 times>, __ss_align = 0}, addr_len = 16, cstate = {readbuf = {<dena::noncopyable> = {<No data fields>}, buffer = 0x5210000dd500 "1\t=\t1\t0\t1\t0\tU\t0\t11\n", '\276' <repeats 181 times>..., begin_offset = 0, end_offset = 19, alloc_size = 4096}, writebuf = {<dena::noncopyable> = {<No data fields>}, buffer = 0x50300006e920 "0\t1\n", '\276' <repeats 28 times>, "\003\021", begin_offset = 0, end_offset = 0, alloc_size = 32}, prep_stmts = std::vector of length 2, capacity 2 = {{dbctx = 0x0, table_id = 18446744073709551615, idxnum = 18446744073709551615, ret_fields = std::vector of length 0, capacity 0, filter_fields = std::vector of length 0, capacity 0}, {dbctx = 0x50e000081140, table_id = 0, idxnum = 0, ret_fields = std::vector of length 2, capacity 2 = {0, 1}, filter_fields = std::vector of length 0, capacity 0}}, resp_begin_pos = 0, find_nl_pos = 0}, err = "", readsize = 0, nonblocking = true, read_finished = false, write_finished = false, nb_last_io = 1729081330, conns_iter = 0x5130000a9fc0, authorized = false}) at /home/dan/repos/mariadb-server-10.5/plugin/handler_socket/handlersocket/hstcpsvr_worker.cpp:672
#18 0x00007fbd1cbdd6dc in dena::hstcpsvr_worker::run_one_ep (this=0x510000009040)
    at /home/dan/repos/mariadb-server-10.5/plugin/handler_socket/handlersocket/hstcpsvr_worker.cpp:556
#19 0x00007fbd1cbdc614 in dena::hstcpsvr_worker::run (this=0x510000009040) at /home/dan/repos/mariadb-server-10.5/plugin/handler_socket/handlersocket/hstcpsvr_worker.cpp:332
#20 0x00007fbd1cbea020 in dena::worker_throbj::operator() (this=<optimized out>) at /home/dan/repos/mariadb-server-10.5/plugin/handler_socket/handlersocket/hstcpsvr.cpp:30
#21 dena::thread<dena::worker_throbj>::thread_main (arg=<optimized out>) at /home/dan/repos/mariadb-server-10.5/plugin/handler_socket/libhsclient/thread.hpp:71
#22 0x00000000007d022d in asan_thread_start(void*) ()
#23 0x00007fbd272a66d7 in start_thread () from /lib64/libc.so.6
#24 0x00007fbd2732a60c in clone3 () from /lib64/libc.so.6
(gdb) frame 11
#11 0x0000000001953f7e in calc_row_difference (uvect=0x520000070cc8, old_row=0x519000221ad8 "", new_row=0x519000221ac8 "", table=0x519000221598, 
    upd_buff=0x50d000020ac8 "\200", buff_len=110, prebuilt=0x520000070108, auto_inc=<optimized out>)
    at /home/dan/repos/mariadb-server-10.5/storage/innobase/handler/ha_innodb.cc:8329
8329		const uint8 op_map= tl->trg_event_map | tl->slave_fk_event_map;
(gdb) p tl
$1 = <optimized out>
(gdb) list
8324		uvect->info_bits = 0;
8325	
8326		ut_a(buf <= (byte*) original_upd_buff + buff_len);
8327	
8328		const TABLE_LIST *tl= table->pos_in_table_list;
8329		const uint8 op_map= tl->trg_event_map | tl->slave_fk_event_map;
8330		/* Used to avoid reading history in FK check on DELETE (see MDEV-16210). */
8331		prebuilt->upd_node->is_delete =
8332			(op_map & trg2bit(TRG_EVENT_DELETE)
8333			 && table->versioned(VERS_TIMESTAMP))
(gdb) p table->pos_in_table_list
$2 = (TABLE_LIST *) 0x7fbceae09180
(gdb) p table->pos_in_table_list->trg_event_map 
$3 = 0 '\000'
(gdb) p table->pos_in_table_list->slave_fk_event_map 
$4 = 0 '\000'
```

dr-m committed:
MDEV-32484: Handlersocket Memory Leak in 10.11

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=3397) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=3397) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: grooverdan<br/>:white_check_mark: bnestere<br/>:x: dr-m<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=3397) it.</sub>
